src/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -424,6 +433,22 @@ internal bool Remove(TKey key)
             }
 
             //----------------------------------------------------------------------------------------
+            // Removes all entries from the table.
+            //
+            // Precondition:
+            //      Must hold _lock.
+            //----------------------------------------------------------------------------------------
+            internal void Clear()
+            {
+                // Remove all handles (as in Remove, just by setting the hashcode and leaving
+                // actual removal up to the finalizer).
+                for (int i = 0; i < _firstFreeEntry; i++)
```

Comment:
> does the handle itself have appropriate synchronization to support reading-while-mutating?

I think so.

> Presumably whatever we do in Clear should also be done in Remove.

Yes.

> it appears to only be used internally via some WinRT functionality in CoreLib

For Clear, it may be actually best to keep ditching the whole container as before. It should be better to get the handles freed sooner rather than later.